share/new_eds/validation.go

```diff
@@ -0,0 +1,64 @@
+package eds
+
+import (
+	"context"
+	"errors"
+	"fmt"
+
+	"github.com/celestiaorg/rsmt2d"
+
+	"github.com/celestiaorg/celestia-node/share"
+	"github.com/celestiaorg/celestia-node/share/shwap"
+)
+
+var _ Accessor = validation{}
+
+// ErrOutOfBounds is returned whenever an index is out of bounds.
+var ErrOutOfBounds = errors.New("index is out of bounds")
+
+// validation is a  Accessor implementation that performs sanity checks on methods. It wraps
+// another  Accessor and performs bounds checks on index arguments.
+type validation struct {
+	Accessor
+}
+
+func WithValidation(f Accessor) Accessor {
+	return &validation{Accessor: f}
+}
+
+func (f validation) Sample(ctx context.Context, rowIdx, colIdx int) (shwap.Sample, error) {
+	if err := validateIndexBounds(ctx, f, colIdx); err != nil {
+		return shwap.Sample{}, fmt.Errorf("col: %w", err)
+	}
+	if err := validateIndexBounds(ctx, f, rowIdx); err != nil {
+		return shwap.Sample{}, fmt.Errorf("row: %w", err)
+	}
```

Comment:
I hope calling Size would be cheap to call it twice

Reply:
added caching for size just in case

Reply:
Cool, just lets make sure it concurrency safe